clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1957,8 +1966,25 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
   }
 
   bool handleStructType(ParmVarDecl *PD, QualType ParamTy) final {
-    Diag.Report(PD->getLocation(), diag::err_bad_kernel_param_type) << ParamTy;
-    IsInvalid = true;
+    CXXRecordDecl *RD = ParamTy->getAsCXXRecordDecl();
+    // For free functions all struct/class kernel arguments are forward declared
+    // in integration header, that adds additional restrictions for kernel
+    // arguments.
+    // Lambdas are not forward declarable. So, diagnose them properly.
+    if (RD->isLambda()) {
+      Diag.Report(PD->getLocation(), diag::err_bad_kernel_param_type)
+          << ParamTy;
+      IsInvalid = true;
+      return isValid();
+    }
+
+    // Check that the type is defined at namespace scope.
+    const DeclContext *DeclCtx = RD->getDeclContext();
+    if (!DeclCtx->isTranslationUnit() && !isa<NamespaceDecl>(DeclCtx)) {
+      Diag.Report(PD->getLocation(), diag::err_bad_kernel_param_type)
+          << ParamTy;
+      IsInvalid = true;
+    }
```

Comment:
I added handling for `LinkageSpecDecl` here, however I'm not able to write the code that would give me `ExternCContextDecl` to handle. Does clang emit it still?

> the forward declaration in the integration header should reproduce the enclosing extern "C", extern "C++", etc... context.

I wonder, what would be the benefit of doing that? I suppose linkage declaration contexts shouldn't affect the name. I see that the code generating forward declarations is intentionally skipping `LinkageSpecDecl` .

Reply:
I looked into `ExternCContextDecl` and confirmed that it is never used as either a lexical or semantic context for a declaration (it is used to collect and identify `extern "C"` declarations that might appear in distinct lexical contexts). So, nothing to be done for it; the change to handle `LinkageSpecDecl` is all that is needed.

> I wonder, what would be the benefit of doing that? I suppose linkage declaration contexts shouldn't affect the name. I see that the code generating forward declarations is intentionally skipping `LinkageSpecDecl` .

It could affect name mangling for variable and function declarations which in turn could affect mangling elsewhere. A highly contrived example involving kernel names is below. It looks like `icx` currently fails to handle such cases regardless of whether `f()` is declared `extern "C"` though;  https://godbolt.org/z/KWhh5xWq9. In general, it looks like `icx` fails to generate correct integration headers for class templates with non-type template parameters that reference other symbols.
```
#include <sycl/sycl.hpp>
extern "C" void f();
template<void(*)()> class kernel_name {};
int main() {
  sycl::queue q;
  q.submit([](sycl::handler &h) {
	h.single_task<kernel_name<f>>([]{});
  });
  q.wait();
}
```

I don't know how important such cases are. Since SYCL doesn't support function pointers in device code, it could be useful to smuggle a function reference through the type system. `icx` accepts the following example: https://godbolt.org/z/TGW5Ynnjz.
```
#include <sycl/sycl.hpp>
extern "C" SYCL_EXTERNAL void f();
template<void(&FN)()> struct X {
  void operator()() const {
	FN();
  }
};
int main() {
  sycl::queue q;
  X<f> x;
  q.submit([=](sycl::handler &h) {
	h.single_task<struct KN>([=]{ x(); });
  });
  q.wait();
}
```

Reply:
Yeah, it seems the problem here is that we don't support forward declaring functions in the integration header. I'm not sure how useful the examples above, since plain SYCL doesn't allow function pointers anyway. I see that supporting these cases may require big functional changes around integration header generation, so I'm not sure if we should do this as a part of the PR. My preference is to add support for these changes as a separate PR. WDYT?